xline/src/storage/index.rs

```diff
@@ -205,72 +197,71 @@ impl IndexOperate for Index {
-                        .map(|pair| (pair, k.clone()))
+                .filter_map(|(entry, i)| {
+                    entry.value().map_write(|mut revs| {
+                        Self::gen_del_revision(
+                            revs.as_mut(),
+                            revision,
+                            sub_revision.overflow_add(i),
+                        )
+                        .map(|pair| (pair, entry.key().clone()))
+                    })
                 })
                 .unzip(),
-            RangeType::Range => inner
-                .index
-                .range_mut(KeyRange::new(key, range_end))
+            RangeType::Range => self
+                .inner
+                .range(KeyRange::new(key, range_end))
                 .zip(0..)
-                .filter_map(|((k, revs), i)| {
-                    Self::gen_del_revision(revs, revision, sub_revision.overflow_add(i))
-                        .map(|pair| (pair, k.clone()))
+                .filter_map(|(entry, i)| {
+                    entry.value().map_write(|mut revs| {
+                        Self::gen_del_revision(
+                            revs.as_mut(),
+                            revision,
+                            sub_revision.overflow_add(i),
+                        )
+                        .map(|pair| (pair, entry.key().clone()))
+                    })
                 })
                 .unzip(),
         };
-        if !keys.is_empty() {
-            assert!(
-                inner
-                    .unavailable_cache
-                    .insert(revision, keys.clone())
-                    .is_none(),
-                "revision {revision} is already in the unavailable cache",
-            );
-        }
         (pairs, keys)
     }
 
-    fn insert_or_update_revision(
-        &self,
-        key: &[u8],
-        revision: i64,
-        sub_revision: i64,
-    ) -> KeyRevision {
-        let mut inner = self.inner.lock();
-        let new_rev = if let Some(revisions) = inner.index.get_mut(key) {
-            if let Some(rev) = revisions.last() {
-                let new_rev = if rev.is_deleted() {
-                    KeyRevision::new(revision, 1, revision, sub_revision)
-                } else {
-                    KeyRevision::new(
-                        rev.create_revision,
-                        rev.version.overflow_add(1),
-                        revision,
-                        sub_revision,
-                    )
-                };
-                revisions.push(new_rev);
-                new_rev
+    fn insert(&self, key_revisions: Vec<(Vec<u8>, KeyRevision)>) {
+        for (key, revision) in key_revisions {
+            if let Some(entry) = self.inner.get::<[u8]>(key.as_ref()) {
+                entry.value().map_write(|mut revs| revs.push(revision));
             } else {
-                panic!("Get empty revision list for key {key:?}");
+                _ = self.inner.insert(key, RwLock::new(vec![revision]));
```

Comment:
Could we use `get_or_insert`?

Reply:
There is no need to do so because even if the `get_or_insert` success succeeds, you still have to check whether the returned value contains the given revision. If the given revision doesn’t exist, you still need to push it manually. Therefore, it won’t save much code.